fabric-transfer-api-v1/src/testmod/java/net/fabricmc/fabric/test/transfer/gametests/VanillaStorageTests.java

```diff
@@ -123,6 +128,85 @@ public void testChestComparator(TestContext context) {
 		context.complete();
 	}
 
+	/**
+	 * Tests that containers such as chests don't update adjacent comparators until the very end of a committed transaction.
+	 */
+	@GameTest(templateName = FabricGameTest.EMPTY_STRUCTURE)
+	public void testChestComparator(TestContext context) {
+		testComparatorOnInventory(context, Blocks.CHEST, ItemVariant.of(Items.DIAMOND));
+	}
+
+	/**
+	 * Same as {@link #testChestComparator} but for chiseled bookshelves, because their implementation is very... strange.
+	 */
+	@GameTest(templateName = FabricGameTest.EMPTY_STRUCTURE)
+	public void testChiseledBookshelfComparator(TestContext context) {
+		testComparatorOnInventory(context, Blocks.CHISELED_BOOKSHELF, ItemVariant.of(Items.BOOK));
+	}
+
+	/**
+	 * Test for chiseled bookshelves, because their implementation is very... strange.
+	 */
+	@GameTest(templateName = FabricGameTest.EMPTY_STRUCTURE)
+	public void testChiseledBookshelf(TestContext context) {
+		ItemVariant book = ItemVariant.of(Items.BOOK);
+
+		BlockPos pos = new BlockPos(0, 1, 0);
+		context.setBlockState(pos, Blocks.CHISELED_BOOKSHELF.getDefaultState());
+		ChiseledBookshelfBlockEntity bookshelf = (ChiseledBookshelfBlockEntity) context.getBlockEntity(pos);
+		InventoryStorage storage = InventoryStorage.of(bookshelf, null);
+
+		// First, check that we can correctly undo insert operations, because vanilla's setStack doesn't permit it without our patches.
+		try (Transaction transaction = Transaction.openOuter()) {
+			if (storage.insert(book, 2, transaction) != 2) throw new GameTestException("Should have inserted 2 books");
+
+			if (bookshelf.getStack(0).getCount() != 1) throw new GameTestException("Bookshelf stack 0 should have size 1");
+			if (!book.matches(bookshelf.getStack(0))) throw new GameTestException("Bookshelf stack 0 should be a book");
+			if (bookshelf.getStack(1).getCount() != 1) throw new GameTestException("Bookshelf stack 1 should have size 1");
+			if (!book.matches(bookshelf.getStack(1))) throw new GameTestException("Bookshelf stack 1 should be a book and quill");
```

Comment:
Doesn't seem like a Book and Quill is ever inserted, nor does this statement check for one, so the exception text is wrong.

Reply:
You're right. I used to check that call with a book and quill, but I removed it. Thanks again.